packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+        // EXCHANGE
+        const exchangeRaw = code[ptr + 1]
+        const n = (exchangeRaw >> 4) + 1
+        const m = (exchangeRaw & 0x0f) + 1
+        if (n + m + 1 > minStackCurrent) {
+          validationError(EOFError.StackUnderflow)
+        }
+      } else if (opcode === 0xec) {
+        // EOFCREATE
+        const target = code[ptr + 1]
+        if (target >= container.header.containerSizes.length) {
+          validationError(EOFError.InvalidEOFCreateTarget)
+        }
+        if (containerTypeMap.has(target)) {
+          if (containerTypeMap.get(target) !== ContainerSectionType.InitCode) {
+            validationError(EOFError.ContainerDoubleType)
+          }
+        }
+        containerTypeMap.set(target, ContainerSectionType.InitCode)
+      } else if (opcode === 0xee) {
+        // RETURNCONTRACT
+
+        if (mode !== ContainerSectionType.InitCode) {
+          validationError(EOFError.ContainerTypeError)
+        }
+
+        const target = code[ptr + 1]
+        if (target >= container.header.containerSizes.length) {
+          validationError(EOFError.InvalidRETURNContractTarget)
+        }
+        if (containerTypeMap.has(target)) {
+          if (containerTypeMap.get(target) !== ContainerSectionType.DeploymentCode) {
+            validationError(EOFError.ContainerDoubleType)
+          }
+        }
+        containerTypeMap.set(target, ContainerSectionType.DeploymentCode)
+      } else if (opcode === 0xd1) {
+        // DATALOADN
+        const dataTarget = readUint16(code, ptr + 1)
+        const endOfSlice = dataTarget + 32
+        if (container.header.dataSize < endOfSlice) {
+          validationError(EOFError.DataLoadNOutOfBounds)
+        }
+      } else if (opcode === 0x00 || opcode === 0xf3) {
+        // STOP / RETURN
+
+        if (mode === ContainerSectionType.InitCode) {
+          validationError(EOFError.ContainerTypeError)
+        }
+      }
+
+      // Move ptr forward over any intermediates (if any)
+      // Note: for EOF this stackDelta is guaranteed to exist
+      const intermediates = stackDelta[opcode].intermediates
+      if (intermediates > 0) {
+        for (let i = 1; i <= intermediates; i++) {
+          addIntermediate(ptr + i)
+        }
+        ptr += intermediates // If the opcode has any intermediates, jump over it
+      }
+      if (ptr >= code.length) {
+        validationError(EOFError.OpcodeIntermediatesOOB)
+      }
+      ptr++ // Move to next opcode
+      if (stackDelta[opcode].terminating === undefined) {
+        // If the opcode is not terminating we can add the next opcode to the reachable opcodes
+        // It can be reached by sequential instruction flow
+        reachableOpcodes.add(ptr)
+
+        // Add next opcode to successorSet
+        // NOTE: these are all opcodes except RJUMP
+        if (opcode !== 0xe0) {
+          successorSet.add(ptr)
+        }
+      }
+
+      // TODO here validate stack / reachability and stack overflow check
```

Comment:
Is this still a TODO?  Feels like we've covered it here.

Reply:
Yes I think you are right!